## 1.Loading and Preprocessing

In [88]:
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.preprocessing import StandardScaler

data=load_breast_cancer()
df = pd.DataFrame(data.data, columns=data.feature_names)
df["Target"] = data.target

df.head()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,Target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [86]:
df.shape

(569, 31)

In [59]:
df.isnull

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
Target                     0
dtype: int64

In [60]:
x = df.drop("Target", axis=1)
y = df["Target"]

In [77]:
# # Feature scaling
# scaler = StandardScaler()
# x_scaled = scaler.fit_transform(x)

# print(x_scaled)

[[ 1.09706398 -2.07333501  1.26993369 ...  2.29607613  2.75062224
   1.93701461]
 [ 1.82982061 -0.35363241  1.68595471 ...  1.0870843  -0.24388967
   0.28118999]
 [ 1.57988811  0.45618695  1.56650313 ...  1.95500035  1.152255
   0.20139121]
 ...
 [ 0.70228425  2.0455738   0.67267578 ...  0.41406869 -1.10454895
  -0.31840916]
 [ 1.83834103  2.33645719  1.98252415 ...  2.28998549  1.91908301
   2.21963528]
 [-1.80840125  1.22179204 -1.81438851 ... -1.74506282 -0.04813821
  -0.75120669]]


### Preprocessing steps performed

#### a.Converted the dataset into a Pandas DataFrame
#### b. checked for misssing values.
#### c. separated target and features.
#### d. Feature scaling

## 2.Classification Algorithm Implementation

In [80]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=42
)
scaler = StandardScaler()

x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [82]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=1000)

model.fit(x_train, y_train)
x_new = pd.DataFrame([[
    17.99,10.38,122.80,1001.0,    
    0.11840,   
    0.27760,  
    0.3001,    
    0.14710,   
    0.2419,    
    0.07871,   

    1.095,    
    0.9053,    
    8.589,     
    153.40,    
    0.006399,  
    0.04904,   
    0.05373,   
    0.01587,   
    0.03003,   
    0.006193,  

    25.38,     
    17.33,    
    184.60,    
    2019.0,    
    0.1622,    
    0.6656,    
    0.7119,    
    0.2654,    
    0.4601,    
    0.11890    
]],
    columns=df.drop("Target",axis=1).columns
     )
x_new = scaler.transform(x_new)

probability = model.predict_proba(x_new)

logistic_prediction= model.predict(x_new)


print("Probability:", probability)
print("Prediction:", logistic_prediction)
print("Prediction:", data.target_names[logistic_prediction])

Probability: [[9.99999994e-01 6.05846547e-09]]
Prediction: [0]
Prediction: ['malignant']


### Logistic Regression predicts the probability that a sample belongs to a particular class

In [83]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(random_state=42)

model.fit(x_train, y_train)

decision_tree_pred = model.predict(x_new)

print("Prediction:", data.target_names[decision_tree_pred])

Prediction: ['malignant']


### Decision Tree makes prediction for a new data by asking a series of YES or NO questions.

In [65]:
#Random Forest Classifier
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(random_state=42)

model.fit(x_train, y_train)

random_forest_pred  = model.predict(x_new)

print("Prediction:", data.target_names[random_forest_pred])


Prediction: ['malignant']


### Random Forest combines multiple Decision Trees and uses their combined predictions to classify the data.

In [66]:
#Support Vector Machine (SVM)
from sklearn.svm import SVC

model = SVC()

model.fit(x_train, y_train)

svm_pred = model.predict(x_new)

print("Prediction:", data.target_names[svm_pred])

Prediction: ['malignant']


### SVM finds a boundary that separates different classes as clearly as possible.

In [67]:
#KNN
from sklearn.neighbors import KNeighborsClassifier

model = KNeighborsClassifier(n_neighbors=5)

model.fit(x_train, y_train)

knn_pred = model.predict(x_new)

print("Prediction:", data.target_names[knn_pred])

Prediction: ['malignant']


### k-NN classifies a new sample based on the classes of its nearest data points. It is suitable because the breast cancer dataset contains numerical features that can be compared based on their distances.

## 3.Model Comparison

In [ ]:
from sklearn.metrics import accuracy_score

In [76]:
# Logistic Regression
model = LogisticRegression(max_iter=1000)
model.fit(x_train, y_train)
logistic_test_pred = model.predict(x_test)

# Decision Tree
model = DecisionTreeClassifier(random_state=42)
model.fit(x_train, y_train)
decision_tree_test_pred = model.predict(x_test)

# Random Forest
model = RandomForestClassifier(random_state=42)
model.fit(x_train, y_train)
random_forest_test_pred = model.predict(x_test)

# SVM
model = SVC()
model.fit(x_train, y_train)
svm_test_pred = model.predict(x_test)

# KNN
model = KNeighborsClassifier(n_neighbors=5)
model.fit(x_train, y_train)
knn_test_pred = model.predict(x_test)

# Accuracy
print("Logistic Regression:", accuracy_score(y_test, logistic_test_pred))
print("Decision Tree:", accuracy_score(y_test, decision_tree_test_pred))
print("Random Forest:", accuracy_score(y_test, random_forest_test_pred))
print("SVM:", accuracy_score(y_test, svm_test_pred))
print("KNN:", accuracy_score(y_test, knn_test_pred))

Logistic Regression: 0.9736842105263158
Decision Tree: 0.9473684210526315
Random Forest: 0.9649122807017544
SVM: 0.9824561403508771
KNN: 0.9473684210526315


### SVM performed the best with an accuracy of 98.25%.
### Decision Tree and KNN performed the worst, both with 94.74% accuracy.